### PR TITLE
Compute shortest distances between given sets of origins and destinations for large diameter graphs

### PR DESCRIPTION
**Pull request by seunghwak**
This code computes shortest distances between given origins and destinations and produces an OD matrix.

This code runs many SSSPs in parallel to extract sufficient parallelism considering that an SSSP run on a large diameter graph has only limited amount of parallelism.

This code uses hash table (cuco::static_map) to reduce memory footprint (and this enable running many SSSPs by cutting memory requirement to store distances from O(V) to the number of vertices around the active frontier).

For small diameter graphs (where one can extract sufficient parallelism with only a single SSSP run), sequentially running multiple SSSPs may be more performant (as hash table access is more expensive than linear array access + additional overheads to run many SSSPs in parallel).

### CODE REVIEW

**Comment by BradReesWork** [comment]
LGTM from cmake perspective

**Review by BradReesWork — APPROVED** [review]

File: cpp/src/prims/vertex_frontier.cuh [hunk 1]
```diff
@@ -204,8 +203,6 @@ class key_bucket_t {
                                           merged_pair_first + merged_vertices.size())),
           handle_ptr_->get_stream());
         merged_tags.resize(merged_vertices.size(), handle_ptr_->get_stream());
-        merged_vertices.shrink_to_fit(handle_ptr_->get_stream());
-        merged_tags.shrink_to_fit(handle_ptr_->get_stream());
```
Comment (ChuckHastings) [inline cpp/src/prims/vertex_frontier.cuh:208]: Is there a reason we're not shrinking these here?

**Review by ChuckHastings — COMMENTED** [review]
Couple of questions.  Otherwise looks good.

File: cpp/include/cugraph/algorithms.hpp [hunk 2]
```diff
@@ -1179,6 +1179,40 @@ void sssp(raft::handle_t const& handle,
...
 
+/*
+ * @brief Compute the shortest distances from the given origins to all the given destinations.
+ *
+ * @throws cugraph::logic_error on erroneous input arguments.
+ *
+ * @tparam vertex_t Type of vertex identifiers. Needs to be an integral type.
+ * @tparam edge_t Type of edge identifiers. Needs to be an integral type.
+ * @tparam weight_t Type of edge weights. Needs to be a floating point type.
+ * @tparam multi_gpu Flag indicating whether template instantiation should target single-GPU (false)
```
[the remark this answers is not in the document]
Comment (seunghwak) [reply cpp/include/cugraph/algorithms.hpp:1190]: Yes.

File: cpp/src/prims/vertex_frontier.cuh (hunk 1 above)
Comment (seunghwak) [reply cpp/src/prims/vertex_frontier.cuh:208]: `shrink_to_fit` basically allocates another array; copies data; and destroy the original array.

The copying part is expensive. If the expected number of duplicates is low, this really does not reduce memory footprint much, increase the peak memory usage (as we temporarily maintain two arrays), and increases execution time. Here two merged arrays typically don't have that many duplicates. I guess paying this cost for few % or less cut in memory is not a good trade-off.

**Review by ChuckHastings — APPROVED** [review]

**Comment by seunghwak** [comment]
@jnke2016 Take a look for this PR especially the code below.

https://github.com/rapidsai/cugraph/pull/3741/files#diff-1d11ed707cec17cd1996e73fe04a22243e3698d052deed45a73244331cbc5286R344

I think addressing this FIXME can be a good first task to work on primitives.

**Review by naimnv — APPROVED** [review]

**Pull request merged**